pydantic/generics.py

```diff
@@ -79,9 +84,14 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
 
 
 def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Type[Any]:
-    if hasattr(type_, '__origin__') and getattr(type_, '__parameters__', None):
-        concrete_type_args = tuple([typevars_map[x] for x in type_.__parameters__])
-        return type_[concrete_type_args]
+    type_args = _get_args(type_)
+    if type_args:
+        concrete_type_args = tuple(resolve_type_hint(arg, typevars_map) for arg in type_args)
+        origin_type = _get_origin(type_) or type_
+        origin_type = _builtin_type_map.get(origin_type, origin_type)
+        return origin_type[concrete_type_args]
```

Comment:
If `type_` is `Type[T]`, then `origin_type=type`. Unfortunately `type` is not subscriptable, so this will fail.
```suggestion
        if origin_type is type:
            return type_[concrete_type_args]
        else:
            return origin_type[concrete_type_args]
```
I'm not sure if this is a problem for any other types, but here's a test case you can use for `Type[T]`:

```
@skip_36
def test_generic_with_generic_type():
    T = TypeVar('T')

    class ModelWithType(GenericModel, Generic[T]):
        some_type: Type[T]

    class ReferenceModel(GenericModel, Generic[T]):
        abstract_base_with_type: ModelWithType[T]

    ReferenceModel[int]
```

Reply:
Thanks! Indeed there are other cases so my "_builtin_type_map" was way to specific.
I added the test and another variant.  python 3.9 will fix this with https://www.python.org/dev/peps/pep-0585/ but I worked around it for python < 3.9 / >= 3.6.